docs/src/content/hardhat3-alpha/learn-more/plugin-migration.md

```diff
@@ -0,0 +1,236 @@
+const plugin: HardhatPlugin = {
+    id: 'plugin-name',
+    tasks: [newTask],
+    ...
+};
+
+export default plugin;
+```
+
+### Task Actions
+
+Inline task action declaration is supported only for development purposes. For production, move your action to a separate file:
+
+```typescript
+// actions/new-task.ts
+export interface TaskActionArguments {
+  taskOption: boolean;
+}
+
+const action = (NewTaskActionFunction<TaskActionArguments> = async (
+  args,
+  hre,
+) => {
+  // task logic
+});
+
+export default action;
+```
+
+Reference it in your task builder:
+
+```typescript
+// tasks/new-task.ts
+import { task } from "hardhat/config";
+import { NewTaskDefinition } from "hardhat/types/tasks";
+
+const newTask: NewTaskDefinition = task("task-name")
+  .setDescription("A Hardhat 3 task")
+  .setAction(import.meta.resolve("./actions/new-task.js"))
+  .addFlag({ name: "taskOption", description: "a boolean CLI flag" })
+  .build();
+
+export default newTask;
+```
+
+## Hooks
+
+Some processes that were previously handled exclusively by tasks might be better served by hooks. Generally task overrides will still work, but here's an example of a post-compilation action which uses a hook instead:
+
+```typescript
+// hook-handlers/solidity.ts
+import type { SolidityHooks } from "hardhat/types/hooks";
+
+export default async (): Promise<Partial<SolidityHooks>> => ({
+  onCleanUpArtifacts: async (context, artifactPaths, next) => {
+    // post-compilation script
+
+    return next(context, artifactPaths);
+  },
+});
+```
+
+Note that the `context` parameter is a subset of the `hre`. Th task runner is excluded, to discourage tight coupling of tasks.
+
+Like tasks, hooks must be registeredon your `HardhatPlugin` object:
+
+```typescript
+// index.ts
+import type { HardhatPlugin } from "hardhat/types/plugins";
+import newTask from "./tasks/new-task.js";
+
+const plugin: HardhatPlugin = {
+  id: "plugin-name",
+  tasks: [newTask],
+  hookHandlers: {
+    solidity: import.meta.resolve("./hook_handlers/solidity.js"),
+  },
+};
+
+export default plugin;
+```
+
+### Config Hooks
+
+The `extendConfig` function no longer exists in Hardhat 3. Instead, use one of the `ConfigHooks`:
+
+```typescript
+// hook-handlers/config.ts
+import type { ConfigHooks } from "hardhat/types/hooks";
+
+export default async (): Promise<Partial<ConfigHooks>> => ({
+  resolveUserConfig: async (userConfig, resolveConfigurationariable, next) => {
+    const resolvedConfig = await next(userConfig, resolveConfigurationVariable);
+
+    const myUserConfig = {
+      ...defaultConfig,
+      ...userConfig.myUserConfig,
+    };
+
+    return {
+      ...resolvedConfig,
+      myUserConfig,
+    };
+  },
+});
+```
+
+## See Also
+
+### Global Options
+
+```typescript
+// index.ts
+import { globalOption } from 'hardhat/config';
+import { ArgumentType } from 'hardhat/types/arguments';
+import type { HardhatPlugin } from 'hardhat/types/plugins';
+
+declare module 'hardhat/types/global-options' {
```

Comment:
Is it important to import the module first?  The examples I've seen have done that, but skipping doesn't cause any problems in my plugins.